transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java

```diff
@@ -89,11 +103,81 @@ public ChannelFuture register(final Channel channel, final ChannelPromise promis
         return promise;
     }
 
+    /**
+     * Adds a task to be run once at the end of next (or current) {@code eventloop} iteration.
+     *
+     * @param task to be added.
+     */
+    final void onEventLoopIteration(Runnable task) {
+        ObjectUtil.checkNotNull(task, "task");
+        if (isShutdown() || isShuttingDown() || isTerminated()) {
+            reject();
+        }
+
+        if (!tailTasks.offer(task)) {
+            reject(task);
+        }
+
+        if (wakesUpForTask(task)) {
+            wakeup(inEventLoop());
+        }
+    }
+
+    /**
+     * Removes a task that was added previously via {@link #onEventLoopIteration(Runnable)}.
+     *
+     * @param task to be removed.
+     *
+     * @return {@code true} if the task was removed as a result of this call.
+     */
+    final boolean removeOnEventLoopIterationTask(Runnable task) {
+        return tailTasks.remove(ObjectUtil.checkNotNull(task, "task"));
+    }
+
     @Override
     protected boolean wakesUpForTask(Runnable task) {
         return !(task instanceof NonWakeupRunnable);
     }
 
+    @Override
+    protected boolean runAllTasks(long timeoutNanos) {
+        assert inEventLoop();
+        runTimeout = timeoutNanos;
+        if (lastRunTimedOut) {
+            //If the last run timedout, then run tail tasks first, otherwise, tail tasks may be starved.
+            lastRunTimedOut = false;
+            if (runAllTasksFrom(tailTasks, timeoutNanos, runTimeoutUpdater) == DeadlineRunResult.TIMED_OUT
+                || runTimeout <= 0) {
+                /**
+                 * Although, it didn't run any actual tasks, returning false, will indicate that there are no tasks
+                 * available, which we are not sure of.
+                 * Note: If every run of tail tasks and normal tasks time out, then this code will alternate, between
+                 * running tail tasks first and at end, which is fair. This is the reason, lastRunTimedOut is not set
+                 * to true here.
+                 */
+                return true;
+            }
+        }
+
+        return super.runAllTasks(runTimeout);
+    }
+
+    @Override
+    protected void afterRunningAllTasks(boolean timedOut) {
+        lastRunTimedOut = timedOut;
+        if (timedOut) {
+            // Do not run tail tasks on time out. Next run of tasks will run the tail tasks first.
+            return;
+        }
+
+        runAllTasksFrom(tailTasks);
+    }
+
+    @Override
+    protected boolean hasTasks() {
+        return super.hasTasks() || !tailTasks.isEmpty();
+    }
+
```

Comment:
if we override `hasTasks()` we should also consider overriding `pendingTasks()`. Also it seems a bit strange if `hasTasks()` returns `true` but then `pollTask()`/`peekTask()`/`takeTask()` returns `null` ... I'm not sure if this will practically happen given the relationship between the two queues but it might be worth commenting somewhere if this is expected behavior?